### PR TITLE
SWATCH-2289: Add billable usage status consumer to update status on records

### PR DESCRIPTION
**Pull request by kflahert**
Jira issue: SWATCH-2289

## Description

Create a consumer for billable-usage.status topic which will update billable_usage_remittance records with status, error_code, and billed_on.

Topic creation PR: https://github.com/RedHatInsights/platform-mq/pull/360

## Testing

IQE Test MR: 

### Setup

1. Insert billable_usage_remittances to be updated:
```
INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'rhel-for-x86-els-payg', 'vCPUs', '2024-04', 'Premium', 'Production', 'azure', 'test;test', 1, '2024-04-23 12:01:57.670879+00', NULL, '5487b805-944f-45ec-b3c9-27f427a6f155', 'AZURE', '80e74ff7-6453-4f4b-976a-1a1b60b9055f', NULL, NULL, NULL);
INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'rhel-for-x86-els-payg', 'vCPUs', '2024-04', 'Premium', 'Production', 'azure', 'test;test', 1, '2024-04-23 11:01:57.670879+00', '2024-04-24 20:58:38.96563+00', '5487b805-944f-45ec-b3c9-27f427a6f155', 'AZURE', '051eb959-a65a-4b67-9402-7ab2af94b82c', NULL, NULL, NULL);
```
1. Start app with:  `DEV_MODE=true ./gradlew :swatch-billable-usage:quarkusDev`

### Steps

1. Produce the following messages to platform.rhsm-subscriptions.billable-usage.status topic:
```
{
  "totalValue": 1.0,
  "windowTimestamp": "2024-05-28T16:00:00Z",
  "aggregateId": "a558aa6d-f951-4ffc-9752-59c8b6901368",
  "aggregateKey": {
    "orgId": "org123",
    "productId": "rhel-for-x86-els-payg",
    "metricId": "vCPUs",
    "sla": "Premium",
    "usage": "Production",
    "billingProvider": "azure",
    "billingAccountId": "test;test"
  },
  "snapshotDates": ["2024-04-23T11:01:57.670879Z"],
  "status": "succeeded",
  "errorCode": null,
  "billedOn": "2024-05-28T11:01:57.670879Z",
  "remittanceUuids": ["80e74ff7-6453-4f4b-976a-1a1b60b9055f"]
}
```
```
{
  "totalValue": 1.0,
  "windowTimestamp": "2024-05-28T16:00:00Z",
  "aggregateId": "a558aa6d-f951-4ffc-9752-59c8b6901368",
  "aggregateKey": {
    "orgId": "org123",
    "productId": "rhel-for-x86-els-payg",
    "metricId": "vCPUs",
    "sla": "Premium",
    "usage": "Production",
    "billingProvider": "azure",
    "billingAccountId": "test;test"
  },
  "snapshotDates": ["2024-04-23T11:01:57.670879Z"],
  "status": "failed",
  "errorCode": "inactive",
  "billedOn": null,
  "remittanceUuids": ["051eb959-a65a-4b67-9402-7ab2af94b82c"]
}
```
### Verification

1. First record should have status=succeeded and billedOn=2024-05-28T11:01:57.670879Z
`select * from billable_usage_remittance where uuid = '80e74ff7-6453-4f4b-976a-1a1b60b9055f'`
1.   Second record should have status=failed and errorCode=inactive:
`select * from billable_usage_remittance where uuid = '051eb959-a65a-4b67-9402-7ab2af94b82c'`

### CODE REVIEW

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+@Slf4j
+@ApplicationScoped
+@AllArgsConstructor
+public class BillableUsageStatusConsumer {
+
+  private final BillableUsageRemittanceRepository remittanceRepository;
+
+  @Incoming(Channels.BILLABLE_USAGE_STATUS)
+  @Blocking
```
Comment (Sgitario) [inline swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java:44]: When using the `@Transactional` annotation, we don't need the `@Blocking` annotation.

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  @Transactional
+  public void process(BillableUsageAggregate billableUsageAggregate) {
+    var remittances = remittanceRepository.findByIdIn(billableUsageAggregate.getRemittanceUuids());
```
Comment (Sgitario) [inline swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java:47]: why not doing an UPDATE statement with the status, billedOn and errorCode to set for a given list of uuids?
This way we perform a single query instead of 1 query to fetch the data and N queries to update the remittance records.

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    remittances.forEach(
+        remittance -> {
+          if (Objects.nonNull(billableUsageAggregate.getStatus())) {
+            remittance.setStatus(
+                RemittanceStatus.fromString(billableUsageAggregate.getStatus().value()));
+          } else {
+            log.error(
+                "Billable Usage Status update received with no status, {}:",
+                billableUsageAggregate);
```
Comment (Sgitario) [inline swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java:56]: is this an expected scenario? If so, we should trace the log just once (before doing the query).

File: swatch-producer-azure/src/main/java/com/redhat/swatch/azure/service/BillableUsageAggregateConsumer.java [hunk 4]
```diff
@@ -169,7 +169,8 @@ private boolean isUsageDateValid(BillableUsageAggregate aggregate) {
     OffsetDateTime startOfCurrentHour =
         OffsetDateTime.now(Clock.systemUTC()).truncatedTo(ChronoUnit.HOURS);
     OffsetDateTime cutoff = startOfCurrentHour.minus(azureUsageWindow);
-    return !aggregate.getWindowTimestamp().isBefore(cutoff);
+    var earliestUsageDate = aggregate.getSnapshotDates().stream().sorted().findFirst();
+    return !earliestUsageDate.map(date -> date.isBefore(cutoff)).orElse(false);
```
Comment (Sgitario) [inline swatch-producer-azure/src/main/java/com/redhat/swatch/azure/service/BillableUsageAggregateConsumer.java:173]: what is the motivation of this change? I don't see any info in the PR description or in the JIRA ticket.

**Review by Sgitario — CHANGES_REQUESTED** [review]
Just a few suggestions and questions.

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java (hunk 2 above)
Comment (kflahert) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java:47]: Makes sense, I'll update!

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java (hunk 3 above)
Comment (kflahert) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageStatusConsumer.java:56]: It's not expected but good point I'll move this before the query.

File: swatch-producer-azure/src/main/java/com/redhat/swatch/azure/service/BillableUsageAggregateConsumer.java (hunk 4 above)
Comment (kflahert) [reply swatch-producer-azure/src/main/java/com/redhat/swatch/azure/service/BillableUsageAggregateConsumer.java:173]: I realized while testing we have an issue here, we need to use snapshot dates and not the aggregate window for this. It belongs in a new card though I'll remove from this PR and make a new card for it.

**Review by Sgitario — APPROVED** [review]
lgtm!

**Comment by kartikshahc** [comment]
@kflahert for my information can you explain what does `windowTimestamp` represent in payload? 
The test steps works fine. Just one thing that I would like to bring up. In production currently there are records with status "pending"  even though it succeeded. Do we want to handle those old records with liquibase script and make it success or null so that there is less confusion?

**Review by kartikshahc — CHANGES_REQUESTED** [review]
Do we need a liquibase change to handle old records?

**Comment by kflahert** [comment]
> @kflahert for my information can you explain what does `windowTimestamp` represent in payload? The test steps works fine. Just one thing that I would like to bring up. In production currently there are records with status "pending" even though it succeeded. Do we want to handle those old records with liquibase script and make it success or null so that there is less confusion?

@kartikshahc windowTimestamp is the bottom of the hour when we get the first billable-usage for an aggregate. This insures we only submit one usage per hour since that should be unique per aggregate. 

As for liquibase I don't think it makes sense to do here since we haven't implemented the status messages producers so we won't update all records if we do it now. I would recommend we do it after https://issues.redhat.com/browse/SWATCH-2290 and https://issues.redhat.com/browse/SWATCH-2291

**Review by kartikshahc — APPROVED** [review]

**Review by kartikshahc — APPROVED** [review]
LGTM

**Comment by Aurobinda55** [comment]
@kflahert  I have inserted two entries mentioned above and checked the log message in swatch-billable-usage-service. But, couldn't see any such message in billable-usage service as you mentioned . And, also I checked the billable_usage_remittance table , it shows  records without any error code . So, just wanted to know , which field in  the setup insert queries  responsible to changes the status succeeded/failed and  from qe side how can we reproduce this?

My insert queries :
```
INSERT INTO public.billable_usage_remittance VALUES (NULL, '14541308', 'rhel-for-x86-els-payg', 'vCPUs', '2024-04', 'Premium', 'Production', 'azure', 'test;test', 1, '2024-04-23 12:01:57.670879+00', NULL, '5487b805-944f-45ec-b3c9-27f427a6f155', 'AZURE', '80e74ff7-6453-4f4b-976a-1a1b60b9055f', NULL, NULL, NULL);
INSERT INTO public.billable_usage_remittance VALUES (NULL, '14541308', 'rhel-for-x86-els-payg', 'vCPUs', '2024-04', 'Premium', 'Production', 'azure', 'test;test', 1, '2024-04-23 11:01:57.670879+00', '2024-04-24 20:58:38.96563+00', '5487b805-944f-45ec-b3c9-27f427a6f155', 'AZURE', '051eb959-a65a-4b67-9402-7ab2af94b82c', NULL, NULL, NULL);

```
remittance record  :

```
swatch-tally-db=# select * from billable_usage_remittance ; account_number |  org_id  |      product_id       | metric_id | accumulation_period |   sla   |   usage    | billing_provider | billing_account_id | remitted_pending_valu
e |    remittance_pending_date    |         retry_after          |               tally_id               | hardware_measurement_type |                 uuid                 
swatch-tally-db=# select * from billable_usage_remittance ;
-[ RECORD 1 ]-------------+-------------------------------------
account_number            | 
org_id                    | 14541308
product_id                | rhel-for-x86-els-payg
metric_id                 | vCPUs
accumulation_period       | 2024-04
sla                       | Premium
usage                     | Production
billing_provider          | azure
billing_account_id        | test;test
remitted_pending_value    | 1
remittance_pending_date   | 2024-04-23 12:01:57.670879+00
retry_after               | 
tally_id                  | 5487b805-944f-45ec-b3c9-27f427a6f155
hardware_measurement_type | AZURE
uuid                      | 80e74ff7-6453-4f4b-976a-1a1b60b9055f
billed_on                 | 
error_code                | 
status                    | 
-[ RECORD 2 ]-------------+-------------------------------------
account_number            | 
org_id                    | 14541308
product_id                | rhel-for-x86-els-payg
metric_id                 | vCPUs
accumulation_period       | 2024-04
sla                       | Premium
usage                     | Production
billing_provider          | azure
billing_account_id        | test;test
remitted_pending_value    | 1
remittance_pending_date   | 2024-04-23 11:01:57.670879+00
retry_after               | 2024-04-24 20:58:38.96563+00
tally_id                  | 5487b805-944f-45ec-b3c9-27f427a6f155
hardware_measurement_type | AZURE
uuid                      | 051eb959-a65a-4b67-9402-7ab2af94b82c
billed_on                 | 
error_code                | 
status                    | 

(END)
```

Remittance data using api  : 

```
In [2]: app.rhsm_subscriptions.get_account_remittance("rhel-for-x86-els-payg")
2024-06-05 06:20:05.501 [    INFO] [root] GET request response url: http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/remittance/accountRemittances?productId=rhel-for-x86-els-payg&orgId=14541308, status_code: 200
2024-06-05 06:20:05.501 [    INFO] [root] Response: [
    {
        "accumulationPeriod": "2024-04",
        "billingAccountId": "test;test",
        "billingProvider": "azure",
        "metricId": "vCPUs",
        "orgId": "14541308",
        "productId": "rhel-for-x86-els-payg",
        "remittanceDate": "2024-04-23T12:01:57.670879Z",
        "remittanceStatus": "null",
        "remittedValue": 2.0
    }
]
```

**Comment by kflahert** [comment]
@Aurobinda55 the only way to test this is to produce messages directly to kafka on the platform.rhsm-subscriptions.billable-usage.status topic until https://issues.redhat.com/browse/SWATCH-2290 and https://issues.redhat.com/browse/SWATCH-2291 are done.
